### DISCUSSION TITLE
Feature request: setCaretPosition()

### DISCUSSION THREAD

**Discussion by thdoan** [FIameCaster/prism-code-editor#13] [Ideas]
It would be so cool if I could do this:

```
const codeEditor = basicEditor(
  '#code-editor',
  {
    language: 'javascript',
    theme: 'vs-code-dark',
  },
  () => codeEditor.setCaretPosition(ln, col) // line, column
);
```

**Comment by jonpyt** [comment]
This exists already. [API Docs for the returned editor](https://prism-code-editor.netlify.app/api/interfaces/index.prismeditor).
The `setSelection` method does something similar to what you want. The only thing is that you need to convert the line and column to an index in the string value. The method is a simple wrapper around [`HTMLTextAreaElement.setSelectionRange()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setSelectionRange), so you might want to read about that too

**Reply by thdoan** [reply]
Thanks, I saw the section on `setSelection()` but in my head I interpreted that setting the text selection, not placing the caret. Cool, I'll play around with this some more.

**Reply by jonpyt** [reply]
Changing the selection and moving the caret, is the same thing really. It just depends on whether the `selectionStart` and `selectionEnd` are the same or different.

**Reply by thdoan** [reply]
Ah, makes sense once you put it that way ^^

**Reply by thdoan** [reply]
Hi, I wasn't able to set the caret using `setSelection`. Can you please let me know what I'm doing wrong?

https://stackblitz.com/edit/vitejs-vite-anepnm?file=src%2Fmain.ts

**Reply by thdoan** [reply]
This finally worked for me 🥳 

```
const codeEditor = basicEditor(
  '#code-editor',
  {
    language: 'javascript',
    theme: 'vs-code-dark',
    value: code,
  },
  // Editor loaded!
  () => setTimeout(() => {
    codeEditor.textarea.focus();
    codeEditor.setSelection(code.length);
  }, 500)
);
```

**Reply by jonpyt** [reply]
Your example on StackBlitz works, the caret is at the end of the text, but the textarea must be focused for the caret to be shown. This is what you did in your last example. You don't need the setTimeout though. I would remove that so the editor is focused immediately instead of after a delay.

**Reply by thdoan** [reply]
Ah, I know why I had to use `setTimeout` -- the editor is being loaded into a modal that has a separate timer to animate a lightbox shadow. The modal inner div (where `#code-editor` lives) is `display:none` until after the the modal is finished opening, then changes to `display:block`. I had to synchronize the `setSelection` and `focus` with this timeout since I guess when there is no layout it can't calculate the position. I've updated the StackBlitz to simulate the modal:

https://stackblitz.com/edit/vitejs-vite-anepnm?file=src%2Fmain.ts

**Reply by jonpyt** [reply]
My solution would be to do something like this:

```javascript
// Simulate opening modal
setTimeout(() => {
  document.querySelector('.modal')?.classList.add('open');
  focusTextarea();
}, 300);

let loaded = false;

const focusTextarea = () => {
  if (!loaded) loaded = true;
  else {
    codeEditor.setSelection(code.length);
    codeEditor.textarea.focus();
  }
}

// Initialize code editor
const code = 'console.log(123);';
const codeEditor = basicEditor(
  '#code-editor',
  {
    language: 'javascript',
    theme: 'vs-code-dark',
    value: code,
  },
  // Editor loaded!
  focusTextarea,
);
```

This way the textarea won't be focused until both the modal is open and the editor is loaded. This example works.

**Reply by thdoan** [reply]
Great solution, I agree -- this is a better way than trying to sync the `setTimeout` delay, which someone else could change in the future also.
